File: sdk-java/sdk/src/main/java/io/probex/ProbexClient.java
package io.probex;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.probex.models.*;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

/**
 * Java client for the PROBEX REST API server.
 *
 * <pre>{@code
 * var client = new ProbexClient("http://localhost:9712");
 * var profile = client.scan(new ScanRequest("https://api.example.com"));
 * var results = client.run(new RunRequest());
 * System.out.println("Passed: " + results.getPassed());
 * }</pre>
 */
public class ProbexClient {

    private final String baseUrl;
    private final HttpClient httpClient;
    private final ObjectMapper mapper;

    public ProbexClient() {
        this("http://localhost:9712");
    }

    public ProbexClient(String baseUrl) {
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();
        this.mapper = new ObjectMapper()
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    /** Check server health. */
    public HealthResponse health() throws ProbexException {
        return doGet("/api/v1/health", HealthResponse.class);
    }

    /** Get the current API profile. */
    public ScanResult getProfile() throws ProbexException {
        return doGet("/api/v1/profile", ScanResult.class);
    }

    /** Scan an API and return the discovered profile. */
    public ScanResult scan(ScanRequest request) throws ProbexException {
        return doPost("/api/v1/scan", request, ScanResult.class);
    }

    /** Run tests and return results. */
    public TestResult run(RunRequest request) throws ProbexException {
        return doPost("/api/v1/run", request, TestResult.class);
    }

    /** Run tests with default options. */
    public TestResult run() throws ProbexException {
        return run(new RunRequest());
    }

    /** Get the latest test results. */
    public TestResult getResults() throws ProbexException {
        return doGet("/api/v1/results", TestResult.class);
    }

    private <T> T doGet(String path, Class<T> responseType) throws ProbexException {
        try {
            var request = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .timeout(Duration.ofSeconds(120))
                    .GET()
                    .build();
            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() >= 400) {
                throw new ProbexException("GET " + path + " returned " + response.statusCode() + ": " + response.body());
            }
            return mapper.readValue(response.body(), responseType);
        } catch (ProbexException e) {
            throw e;
        } catch (IOException | InterruptedException e) {
            throw new ProbexException("GET " + path + " failed: " + e.getMessage(), e);
        }
    }

    private <T> T doPost(String path, Object body, Class<T> responseType) throws ProbexException {
        try {
            var json = mapper.writeValueAsString(body);
            var request = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .timeout(Duration.ofSeconds(120))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(json))
                    .build();
            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() >= 400) {
                throw new ProbexException("POST " + path + " returned " + response.statusCode() + ": " + response.body());
            }
            return mapper.readValue(response.body(), responseType);
        } catch (ProbexException e) {
            throw e;
        } catch (IOException | InterruptedException e) {
            throw new ProbexException("POST " + path + " failed: " + e.getMessage(), e);
        }
    }
}

File: sdk-kotlin/src/main/kotlin/io/probex/ProbexClient.kt
package io.probex

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import io.probex.models.*
import kotlinx.serialization.json.Json

/**
 * Kotlin coroutine-based client for the PROBEX REST API.
 *
 * ```kotlin
 * val client = ProbexClient("http://localhost:9712")
 * val profile = client.scan(ScanRequest("https://api.example.com"))
 * val results = client.run()
 * println("Passed: ${results.passed}")
 * client.close()
 * ```
 */
class ProbexClient(
    baseUrl: String = "http://localhost:9712",
    private val httpClient: HttpClient = defaultClient(),
) : AutoCloseable {

    private val baseUrl: String = baseUrl.trimEnd('/')

    companion object {
        private fun defaultClient(): HttpClient = HttpClient(CIO) {
            install(ContentNegotiation) {
                json(Json {
                    ignoreUnknownKeys = true
                    isLenient = true
                })
            }
            engine {
                requestTimeout = 120_000
            }
        }
    }

    /** Check server health. */
    suspend fun health(): HealthResponse = doGet("/api/v1/health")

    /** Get the current API profile. */
    suspend fun getProfile(): ScanResult = doGet("/api/v1/profile")

    /** Scan an API target. */
    suspend fun scan(request: ScanRequest): ScanResult = doPost("/api/v1/scan", request)

    /** Run tests with options. */
    suspend fun run(request: RunRequest = RunRequest()): TestResult =
        doPost("/api/v1/run", request)

    /** Get the latest test results. */
    suspend fun getResults(): TestResult = doGet("/api/v1/results")

    // --- AI endpoints ---

    /** Check AI brain health. */
    suspend fun aiHealth(): AIHealthResponse = doGet("/api/v1/ai/health")

    /** Generate AI-powered test scenarios. */
    suspend fun aiScenarios(request: ScenarioRequest): ScenarioResponse =
        doPost("/api/v1/ai/scenarios", request)

    /** Generate tests from natural language description. */
    suspend fun aiNLToTest(request: NLTestRequest): NLTestResponse =
        doPost("/api/v1/ai/nl-to-test", request)

    private suspend inline fun <reified T> doGet(path: String): T {
        val response = httpClient.get("$baseUrl$path")
        if (!response.status.isSuccess()) {
            val body = response.bodyAsText()
            throw ProbexException("GET $path returned ${response.status.value}: $body")
        }
        return response.body()
    }

    private suspend inline fun <reified R, reified T> doPost(path: String, body: R): T {
        val response = httpClient.post("$baseUrl$path") {
            contentType(ContentType.Application.Json)
            setBody(body)
        }
        if (!response.status.isSuccess()) {
            val text = response.bodyAsText()
            throw ProbexException("POST $path returned ${response.status.value}: $text")
        }
        return response.body()
    }

    override fun close() {
        httpClient.close()
    }
}
